src/devicecontroller/DefaultDeviceController.ts

```diff
@@ -462,48 +621,127 @@ export default class DefaultDeviceController implements DeviceControllerBasedMed
     return navigator.mediaDevices.getDisplayMedia(streamConstraints);
   }
 
-  releaseMediaStream(mediaStreamToRelease: MediaStream | null): void {
-    if (!mediaStreamToRelease) {
-      return;
-    }
-    let tracksToStop: MediaStreamTrack[] | null = null;
+  /**
+   * This function helps `releaseMediaStream` do the right thing.
+   *
+   * We need to do three things:
+   *
+   * * Close the tracks of the source stream.
+   * * Remove the transform.
+   * * Clean up the _source_ stream's callback, as if `releaseMediaStream` had
+   *   been called with that stream -- that's the stream that's tracked in
+   *   `activeDevices` and needs to have its callbacks removed.
+   *
+   * This is a little fiddly because the stream broker interface doesn't
+   * know about the innards of the device controller, and only has the
+   * meeting session state's stream to work with.
+   *
+   */
+  releaseAudioTransformStream(): void {
+    this.logger.info('Stopping audio track for Web Audio graph');
 
-    if (
-      !!this.audioInputDestinationNode &&
-      mediaStreamToRelease === this.audioInputDestinationNode.stream
-    ) {
-      // release the true audio stream if WebAudio is used.
-      this.logger.info('stopping audio track');
-      tracksToStop = this.audioInputSourceNode.mediaStream.getTracks();
-      this.audioInputSourceNode.disconnect();
-    } else {
-      tracksToStop = mediaStreamToRelease.getTracks();
+    /* istanbul ignore next */
+    const audioEnded = this.activeDevices['audio']?.endedCallback;
+    for (const track of this.audioInputSourceNode.mediaStream.getTracks()) {
+      track.stop();
+
+      /* istanbul ignore else */
+      if (audioEnded) {
+        track.removeEventListener('ended', audioEnded);
+        delete this.activeDevices['audio'].endedCallback;
+        delete this.activeDevices['audio'];
```

Comment:
Just curious. What happen if we just delete `this.activeDevices['audio']` and not the `endedCallback`?

Reply:
If the previous value of `this.activeDevices['audio']` were still present anywhere — including in console logs, debugger-stored variables, etc. — then the callback would leak. Deleting the callback from the device itself helps to aggressively prune the object graph.

This is also a piece of code that is directly related to cleaning up references to this callback, so it's nice to be explicit about where the leaks might lie.

Reply:
Got it